smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,365 @@
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import dataclasses
+import subprocess as sp
+import typing as t
+import uuid
+
+from typing_extensions import Self, TypeAlias, TypeVarTuple, Unpack
+
+from smartsim._core.utils import helpers
+from smartsim.error import errors
+from smartsim.types import LaunchedJobID
+
+if t.TYPE_CHECKING:
+    from smartsim.experiment import Experiment
+    from smartsim.settings.arguments import LaunchArguments
+
+_Ts = TypeVarTuple("_Ts")
+_T_contra = t.TypeVar("_T_contra", contravariant=True)
+
+_DispatchableT = t.TypeVar("_DispatchableT", bound="LaunchArguments")
+"""Any type of luanch arguments, typically used when the type bound by the type
+argument is a key a `Dispatcher` dispatch registry
+"""
+_LaunchableT = t.TypeVar("_LaunchableT")
+"""Any type, typically used to bind to a type accepted as the input parameter
+to the to the `LauncherProtocol.start` method
+"""
+
+_EnvironMappingType: TypeAlias = t.Mapping[str, "str | None"]
+"""A mapping of user provided mapping of environment variables in which to run
+a job
+"""
+_FormatterType: TypeAlias = t.Callable[
+    [_DispatchableT, "ExecutableProtocol", _EnvironMappingType], _LaunchableT
+]
+"""A callable that is capable of formatting the components of a job into a type
+capable of being launched by a launcher.
+"""
+_LaunchConfigType: TypeAlias = (
+    "_LauncherAdapter[ExecutableProtocol, _EnvironMappingType]"
+)
+"""A launcher adapater that has configured a launcher to launch the components
+of a job with some pre-determined launch settings
+"""
+_UnkownType: TypeAlias = t.NoReturn
+"""A type alias for a bottom type. Use this to inform a user that the parameter
+a parameter should never be set or a callable will never return
+"""
+
+
+@t.final
+class Dispatcher:
+    """A class capable of deciding which launcher type should be used to launch
+    a given settings type.
+    """
+
+    def __init__(
+        self,
+        *,
+        dispatch_registry: (
+            t.Mapping[type[LaunchArguments], _DispatchRegistration[t.Any, t.Any]] | None
+        ) = None,
+    ) -> None:
+        """Initialize a new `Dispatcher`
+
+        :param dispatch_registry: A pre-configured dispatch registry that the
```

Comment:
This dispatch registry is created at program compile time? And uses the @dispatch ? If so do you mind adding this to the param desc

Reply:
Of course!!

Reply:
Added clarification at the class level docstring instead explaining the usage of `Dispatcher.get_dispatch` and `Dispatcher.dispatch`, including the fact that `Dispatcher.dispatch` can be used as a decorator.